### PR TITLE
[monitoring] Rethink when check threshold #223

### LINKED ISSUE #223: [monitoring] Rethink when it's best to check threshold (eg: after metric is written)
We recently changed the metric write operation to be run in the background #190, but the threshold is checked right after the write operation is sent to be performed to the background worker:

https://github.com/openwisp/openwisp-monitoring/blob/f6c75913ecadf8d17e4ba3d1a9c248cf11ff5689/openwisp_monitoring/monitoring/base/models.py#L207-L220

The `check_threshold` method handles a case in which past data is used for the check, but the code has not been updated after the changes to #190, so several inconsistencies may appear:

- the check threshold operation may be performed before the value is written to the tsdb, which can cause inconsistencies in behavior since the code has been written to be executed after the values have been written
- the check threshold may fail if in that moment the tsdb is experiencing issues

Taking this into account we should think about a solution.
I think we should find a way to ensure the check threshold operation is done after the metric has been written successfully to the tsdb.

### PR DESCRIPTION
**Pull request by nepython**
Closes #223

Checks:

- [x] I have manually tested the proposed changes
- [x] I have written new test cases to avoid regressions (if necessary)
- [x] I have updated the documentation (e.g. README.rst)

### CODE REVIEW

File: openwisp_monitoring/monitoring/tests/test_monitoring_notifications.py
Comment (nemesisdesign) [inline openwisp_monitoring/monitoring/tests/test_monitoring_notifications.py:175]: during tests celery tasks are executed synchronously: https://github.com/openwisp/openwisp-monitoring/blob/3d55db0a94933515ce410989c9e8c2ca24b61088/tests/openwisp2/settings.py#L165-L170

Therefore this sleep should not be needed.

File: openwisp_monitoring/monitoring/base/models.py
Comment (nemesisdesign) [inline openwisp_monitoring/monitoring/base/models.py:227]: I think now is a good time to move this signal to the background task and execute it after the metric has been written.
We should also update the related docs to explain it will fire only after a successful metric write.
The way it's implemented now is really not very useful.

File: openwisp_monitoring/monitoring/tasks.py
Comment (nemesisdesign) [inline openwisp_monitoring/monitoring/tasks.py:22]: the try/except block should include only this line:

```python
try:
    DB query
except ObjectDoesNotExist:
    return
```

File: openwisp_monitoring/monitoring/base/models.py
Comment (nemesisdesign) [inline openwisp_monitoring/monitoring/base/models.py:219]: `threshold_dict` should be named `check_threshold_kwargs`, which indicates these are the keyword arguments for the `check_threshold()` method.

File: openwisp_monitoring/monitoring/tasks.py
Comment (nemesisdesign) [inline openwisp_monitoring/monitoring/tasks.py:30]: is there a simpler way to do this? Isn't the time we get always supposed to be in UTC?

Also, I think it would be better to move it to `check_threshold`
I wold simply call `metric.check_threshold(**check_threshold_kwargs)` and let `check_threshold` figure out what it has to do.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Good progress, see my comments below. I have not tested this yet.

File: openwisp_monitoring/monitoring/base/models.py
Comment (nepython) [reply openwisp_monitoring/monitoring/base/models.py:227]: Agreed :+1:

File: openwisp_monitoring/monitoring/base/models.py
Comment (nemesisdesign) [inline openwisp_monitoring/monitoring/base/models.py:159]: the only potentially dangerous line I see is this one.

**Review by nemesisdesign — COMMENTED** [review]
Looks good, I need to find some time to test it manually to ensure everything keeps working normally.

File: openwisp_monitoring/monitoring/base/models.py
Comment (nepython) [reply openwisp_monitoring/monitoring/base/models.py:159]: To be honest, I agree with you. I feel slightly uncomfortable with timezones and their implications (as to where UTC is to be used and where the local timezone) right now. What I am sure of is that this datetime object needed to be time aware, `Z` implied UTC timezone hence I went with it. Please let me know if I am misunderstanding this anywhere or if you meant dangerous in a different sense, thanks.

File: requirements.txt
Comment (nemesisdesign) [inline requirements.txt:10]: It's possible that the library is already being pulled because it's a dependency of some other package, but I added it anyway with a relaxed version range to follow the **"explicit is better than implicit"** approach.
The version range is relaxed because it's a very stable library, at the same time we won't accept an eventual 3.0 automatically without manual testing.

**Review by nemesisdesign — COMMENTED** [review]
I changed the date parsing code, see the commit message for an explanation:
https://github.com/openwisp/openwisp-monitoring/pull/246/commits/c05ac7fd40af27cdba9e2576ab70494c33ff1e4a

Tested manually, haven't found issues :+1:.

Let's wait for the build to finish, once it's successful we can merge.

File: requirements.txt
Comment (nepython) [reply requirements.txt:10]: Understood the approach from the commit message and it seems much more readable and simple. I didn't use this library earlier, thanks for introducing an application for the same :)

**Review by nemesisdesign — APPROVED** [review]

**Pull request merged**
